subprojects/GOPATH/src/github.com/microsoft/hivedscheduler/pkg/algorithm/hived_algorithm.go

```diff
@@ -275,87 +328,291 @@ func (h *HivedAlgorithm) GetAffinityGroup(name string) api.AffinityGroup {
 		name)))
 }
 
+func (h *HivedAlgorithm) GetClusterStatus() api.ClusterStatus {
+	h.algorithmLock.RLock()
+	defer h.algorithmLock.RUnlock()
+
+	s := api.ClusterStatus{
+		PhysicalCluster: h.apiClusterStatus.PhysicalCluster.DeepCopy(),
+		VirtualClusters: map[api.VirtualClusterName]api.VirtualClusterStatus{},
+	}
+	for vcn, vcs := range h.apiClusterStatus.VirtualClusters {
+		s.VirtualClusters[vcn] = vcs.DeepCopy()
+	}
+	return s
+}
+
+func (h *HivedAlgorithm) GetPCStatus() api.PhysicalClusterStatus {
+	h.algorithmLock.RLock()
+	defer h.algorithmLock.RUnlock()
+
+	return h.apiClusterStatus.PhysicalCluster.DeepCopy()
+}
+
+func (h *HivedAlgorithm) GetAllVCsStatus() map[api.VirtualClusterName]api.VirtualClusterStatus {
+	h.algorithmLock.RLock()
+	defer h.algorithmLock.RUnlock()
+
+	allVcs := map[api.VirtualClusterName]api.VirtualClusterStatus{}
+	for vcn, vcs := range h.apiClusterStatus.VirtualClusters {
+		allVcs[vcn] = vcs.DeepCopy()
+	}
+	return allVcs
+}
+
+func (h *HivedAlgorithm) GetVCStatus(vcn api.VirtualClusterName) api.VirtualClusterStatus {
+	h.algorithmLock.RLock()
+	defer h.algorithmLock.RUnlock()
+
+	return h.apiClusterStatus.VirtualClusters[vcn].DeepCopy()
+}
+
 // validateInitialAssignment makes sure that the initial cell assignments
 // to all VCs can be fit into the configured physical cells.
+// This method also initiates the totalLeftCellNum and badFreeCellNum structures.
 func (h *HivedAlgorithm) validateInitialAssignment() {
-	totalQuota := map[CellChain]map[CellLevel]int32{}
-	for vc, vcs := range h.vcSchedulers {
-		for chain, ccl := range vcs.getNonReservedCellList() {
-			if totalQuota[chain] == nil {
-				totalQuota[chain] = map[CellLevel]int32{}
+
+	h.allVCFreeCellNum = map[CellChain]map[CellLevel]int32{}
+	for _, vcQuota := range h.vcFreeCellNum {
+		for chain, chainQuota := range vcQuota {
```

Comment:
>chainQuota [](start = 13, length = 10)

Rename quota to be a more concert concept. Such as total/free cells. Also for all places.